### PR TITLE
Azautogen docs

### LINKED ISSUE #551: [ISSUE][DOCS] Missing docs for AzAutoGen
## Describe the issue

The existing documentation for the AzAutoGen system (nee: AzCodeGen) is incredibly out of date and describes a system no longer in use. This section needs to be dumped and rewritten, and has the advantage that the AzAutoGen replacement system is dirt-simple, has lots of internal documentation already, and has good examples in the source code as-is.

## Impacted page(s)

https://o3de.org/docs/user-guide/engine/autogen/

## Expected experience

Accurate and up to date information on AzAutoGen including:

* An overview
* An example of basic usage
* A reference for command-line options

### PR DESCRIPTION
**Pull request by sptramer**
## Change summary

Add documentation for the AzAutoGen source input generator (resolves #551)

### Submission Checklist:

* [x] **Descriptive active voice** - Do descriptive sentences have a clear *subject* and *action verb*?
* [x] **Answer the question at hand** - Does the documentation answer a *what*, *why*, *how*, or *where* type of question?
* [x] **Consistency** - Does the content consistently follow the [style guide](https://o3de.org/docs/contributing/to-docs/style-guide/)?
* [x] **Help the user** - Does the documentation show the user something *meaningful*?

### CODE REVIEW

File: content/docs/user-guide/programming/autogen/_index.md [hunk 1]
```diff
@@ -1,10 +1,237 @@
...
+
+In this topic you'll be introduced to the conceptual idea behind AzAutoGen, and become familiar with both of the JSON and XML input formats. To learn more about Jinja templates, read the [Jinja template designer](https://jinja.palletsprojects.com/templates/) documentation.
+
+## How AzAutoGen works
+
+AzAutoGen works off of two input sets, a collection of value and template files, and a series of mappings from values to templates. AzAutoGen outputs to a specified file, or a set of files which contain special substitutions where When AzAutoGen is invoked, it runs through the following steps:
```
Comment (willihay) [inline content/docs/user-guide/programming/autogen/_index.md:14]:
```suggestion
AzAutoGen works off of two input sets, a collection of value and template files, and a series of mappings from values to templates. AzAutoGen outputs to a specified file, or a set of files which contain special substitutions. When AzAutoGen is invoked, it runs through the following steps:
```

File: content/docs/user-guide/programming/autogen/_index.md [hunk 2]
```diff
@@ -1,10 +1,237 @@
...
+
+1. Prune all input files without one of the extensions `.xml`, `.json`, or `.jinja`. Templates can be co-located with other code as a result, although this isn't recommended.
+1. Categorize the input files as source files (`.xml`, `.json`) and template files (`.jinja`).
+1. For each `input_map,template,output` tuple that makes up a rule, match the corresponding files into their appropriate sets. `template` must be a single file; Each input is run through this template to generate a corresponding file named according to the pattern in `output`. Each mapping argument allows for the following substitutions:
+
+   * Within `input_map`, the following matching operators are supported.
+     * `*` - Sequence of any length
+     * `?` - Single-character squence
```
Comment (willihay) [inline content/docs/user-guide/programming/autogen/_index.md:22]:
```suggestion
     * `*` - Sequence of any length.
     * `?` - Single-character sequence.
```

File: content/docs/user-guide/programming/autogen/_index.md [hunk 3]
```diff
@@ -1,10 +1,237 @@
...
+     * `[<sequence>]` - Matches any characters in `<sequence>`.
+     * `[!<sequence>]` - Matches any characters _not_ in `<sequence>`.
+   * Within `output`, the following tokens can be used to create unique file names for multiple inputs.
+      * `$path`  - The path to the final output destination.
+      * `$fileprefix` - The name of the current file up to the first `.` token. This is equivalent to running the UNIX shell command `basename ${file%%.*}`.
+      * `$file` - The full name of the input file. This is equivalent to running the UNIX shell command `basename $file`.
+
+   1. For each input file:
```
Comment (willihay) [inline content/docs/user-guide/programming/autogen/_index.md:30]: Did you mean to indent this? It feels like it could be step 4 after the first 3 steps, but it's not because of the indentation.

File: content/docs/user-guide/programming/autogen/_index.md [hunk 4]
```diff
@@ -1,10 +1,237 @@
...
+      1. The input is processed through a single Jinja template. Inside of the template, Jinja has access to a native Python object representing the contents of the input file.
+      2. The output is written to its corresponding file.
+
+<!-- TODO: Add a simple flowchart. -->
```
Comment (willihay) [inline content/docs/user-guide/programming/autogen/_index.md:34]: Create an issue for this so the TODO doesn't get forgotten.

File: content/docs/user-guide/programming/autogen/_index.md [hunk 5]
```diff
@@ -1,10 +1,237 @@
...
+            "Type": "uint8_t",
+            "Name": "chunkCount",
+            "Init": "0"
+          },{
+            "Type": "AzNetworking::ChunkBuffer",
+            "Name": "chunkBuffer"
+        }]
+      }
+    ]
+  }
```
Comment (willihay) [inline content/docs/user-guide/programming/autogen/_index.md:216]: One too many closing braces...
```suggestion
```

File: content/docs/user-guide/programming/autogen/_index.md [hunk 6]
```diff
@@ -1,10 +1,237 @@
...
+description: Use the Open 3D Engine (O3DE) Automated Generator to generate resources from Jinja2 templates. 
+linktitle: AzAutoGen source generator
+title: Automate source generation from templates with AzAutoGen
 weight: 200
 ---
 
-{{< todo issue="https://github.com/o3de/o3de.org/issues/551" >}}
-Write documentation for the AZ Generator system.
-{{< /todo >}}
+When confronted with the problem of many similar code objects or source inputs, the best solution for simple cases is often automated generation from templates and values. **Open 3D Engine (O3DE)** has its own lightweight generator system, *AzAutoGen*, to solve this problem. AzAutoGen uses the [Jinja](https://jinja.palletsprojects.com/) template engine and Python in combination with either XML or JSON input files to produce generated sets of output. Your own O3DE Gems and projects can take advantage of AzAutoGen during [build](/docs/user-guide/build/generated-source/) to generate source inputs.
```
Comment (willihay) [inline content/docs/user-guide/programming/autogen/_index.md:8]: "When confronted with the problem of many similar code objects or source inputs, [...]"

I don't feel like I understand the problem here. The word choice makes the problem domain too ambiguous. Can you rephrase, and follow with an example?

**Review by willihay — CHANGES_REQUESTED** [review]

File: content/docs/user-guide/appendix/glossary.md [hunk 7]
```diff
@@ -28,6 +28,9 @@ Any art, texture, 3D model, sound effect, or other digital data that is presente
 ### attachment
 A hierarchical object that is attached to characters, respond to real-world physics, and can be attached, detached, or replaced at runtime in the game. Character attachments include clothing, weapons, tools, or entire body parts such as heads or hands.
 
+### AzAutoGen
+The tool used to create generated source files from input templates and XML or JSON data files. AzAutoGen is meant for use as a simple transpiler when creating large numbers of boilerplate classes or assets.
```
Comment (AMZN-Gene) [inline content/docs/user-guide/appendix/glossary.md:32]: yay! i learned a new word "transpiler"...

**Review by AMZN-Gene — APPROVED** [review]

**Review by paucom — CHANGES_REQUESTED** [review]
Editorial feedback

**Comment by sptramer** [comment]
Closing as stale. Would like to pick it back up someday!

File: content/docs/user-guide/appendix/glossary.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/appendix/glossary.md:32]: Can drop 'many' in this description as well I would guess? Quantity isn't really any kind of requirement here.
```
The tool used to generate source files from input templates and XML or JSON data files. AzAutoGen is meant for use as a simple transpiler when creating boilerplate classes or assets.
```

File: content/docs/user-guide/build/generated-source.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
+---
+linktitle: Build Generated Source Files
+title: Build Generated Source Files for Open 3D Engine
+description: >-
+    Learn how to create template files and their inputs using the Open 3D Engine AzAutoGen generator, and then tie them to build
+    tasks in CMake for inclusion in your target.
```
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/build/generated-source.md:6]: It's an automatic programming mechanism, 
https://en.wikipedia.org/wiki/Automatic_programming

File: content/docs/user-guide/build/generated-source.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+---
+
+For complicated systems which use a lot of boilerplate code, it's useful to perform some kind of light automation to generate source inputs where possible. During builds, **Open 3D Engine (O3DE)** uses the [AzAutoGen](/docs/user-guide/programming/autogen/) tool through a CMake function in order to perform code generation. This emitted source code is stored in your CMake build directory, and picked up to be built in your target.
```
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/build/generated-source.md:9]: `source files` would probably read better than `source inputs`? The generated files are outputs of the AzAutoGen system, and typically inputs to the C++ compiler although that is not required.
`light automation to generate source files where possible`

File: content/docs/user-guide/build/generated-source.md [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+The following special values can be used in an AzAutoGen invocation during build:
+
+* `$path`  - The path to the final output destination.
+* `$fileprefix` - The name of the current file up to the first `.` token. This is equivalent to running the UNIX shell command `basename ${file%%.*}`.
+* `$file` - The full name of the input file. This is equivalent to running the UNIX shell command `basename $file`.
+
+{{< important >}}
+When generating multiple output files, the output name should _always_ begin with `$path`. Otherwise output files may be placed in an incorrect directory.
+{{< /important >}}
```
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/build/generated-source.md:23]: Agreed with having a simple example preceding the glossary of wildcards. That would provide a lot more context to those reading the docs.

File: content/docs/user-guide/build/generated-source.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    AUTOGEN_RULES
+        *.AutoPackets.xml,AutoPackets_Header.jinja,$path/$fileprefix.AutoPackets.h
+        *.AutoPackets.xml,AutoPackets_Inline.jinja,$path/$fileprefix.AutoPackets.inl
+        *.AutoPackets.xml,AutoPackets_Source.jinja,$path/$fileprefix.AutoPackets.cpp
+)
+```
+
+### Integrating with any target
+
+To run AzAutoGen when building a target which isn't defined by `ly_add_target()`, any build target can have an autogen step added by calling the `ly_add_autogen()` CMake function. This function associates a target with a set of autogen rules, and gives an explicit list out the output files generated which should be included in the target build.
```
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/build/generated-source.md:47]: No idea, .. but I guess it doesn't hurt to include as long as it's clear it's not required for normal usage of the feature.

File: content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/overview.md [hunk 12]
```diff
@@ -59,9 +59,17 @@ Since each entity with a multiplayer component is replicated across the network,
...
-* **Autonomous**: A role with the _illusion_ of write access. Autonomous roles are usually assigned to components directly under local user control, and receive a larger amount of network information than a Client role. Autonomous roles can also take advantage of predictive networking.
+* **Autonomous**: A role with the _illusion_ of write access. Autonomous roles are usually assigned to components directly under local user control, and receive a larger amount of network information than a Client role. Autonomous roles can also take advantage of predictive networking.
 * **Authority**: This role has ultimate authority. It has full write access, and sees all network properties on the component.
 
 In addition to the roles described above, O3DE has an additional role found only in multiserver instances:
 
 * **ServerProxy**: Provided to servers which *don't* have authority over a particular component. The proxy dummies out the networking properties which the server should treat as if they were in the Client role. In the event of a server crash, proxy servers can fully assume the Server role as a hot backup.
```
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/overview.md:67]: It's simply a `Server` role now, not `ServerProxy`, to provide symmetry with the `Client` role. The enumeration values are explicit, `NetEntityRole::Client, NetEntityRole::Autonomous, NetEntityRole::Server, NetEntityRole::Authority`

[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/overview.md:67]: In the event of a crash, a `Server` role can assume an `Authority` role, _however_ we should strike that from these docs, as we don't actually ship that functionality with O3DE itself.

File: content/docs/user-guide/programming/autogen/_index.md (hunk 6 above)
Comment (kberg-amzn) [reply content/docs/user-guide/programming/autogen/_index.md:8]: Again, source files would probably read better than source inputs, as the generated files are outputs of the AzAutoGen system. Referring to them as source inputs could cause confusion.

File: content/docs/user-guide/programming/autogen/_index.md [hunk 13]
```diff
@@ -1,10 +1,236 @@
...
+
+In this topic you'll be introduced to the conceptual idea behind AzAutoGen, and become familiar with both of the JSON and XML input formats. To learn more about Jinja templates, read the [Jinja template designer](https://jinja.palletsprojects.com/templates/) documentation.
+
+## How AzAutoGen works
+
+AzAutoGen works off of two input sets, a collection of value and template files, and a series of mappings from values to templates. AzAutoGen outputs to a specified file, or a set of files which contain special substitutions. When AzAutoGen is invoked, it runs through the following steps:
```
[the remark this answers is not in the document]
Comment (kberg-amzn) [reply content/docs/user-guide/programming/autogen/_index.md:14]: Value and template files is a bit awkward.. templates are fine, but consider using `data files` instead of `value`?

File: content/docs/user-guide/build/generated-source.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (chanmosq) [reply content/docs/user-guide/build/generated-source.md:47]: @kberg-amzn Does this mean that users will often use `ly_add_target(...AUTOGEN_RULES...)` -- however use of `ly_add_target()` also exists, but isn't common.  

If this method isn't used, and we don't have a good use case/example for when to use it, keeping this section may cause confusion.

File: content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/overview.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (chanmosq) [reply content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/overview.md:67]: Removed in the new PR for AzAutoGen docs
